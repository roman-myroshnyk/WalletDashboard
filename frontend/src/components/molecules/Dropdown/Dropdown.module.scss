@import '@/styles/colors.scss';

.select {
  position: relative;

  display: inline-block;
  box-sizing: 'border-box';
  margin: 0;
  padding: 15px 50px 15px 15px;

  color: $gray;
  font-size: 16px;

  background-color: $darkBlack;
  background-repeat: no-repeat;
  background-position: right 15px center;
  border: none;
  border-radius: 8px;
  outline: 0;
  outline-width: 0;
  box-shadow: inset 0 0 0 2px $lightBlack;
  cursor: pointer;
  opacity: 1;

  appearance: none;
  pointer-events: auto;
  caret-color: $gray;
}

.select:focus {
  color: $gray;

  background-color: $darkBlack;
  box-shadow: inset 0 0 0 1px $blue;
}

.select:disabled {
  color: $darkGray;

  box-shadow: inset 0 0 0 0 $blue;
  cursor: default;
  opacity: 0.6;
  caret-color: $darkGray !important;
}
