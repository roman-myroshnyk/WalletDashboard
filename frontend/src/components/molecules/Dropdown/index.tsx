/* eslint-disable react/no-array-index-key */
import {
  FC,
  useState,
  useEffect,
  useCallback,
  ChangeEvent,
} from 'react';
import styles from './Dropdown.module.scss';

export interface IProps {
  options?:string[];
  value?: number;
  disabled?:boolean;
  onChange?: (index:number, label:string) => void;
  selectTestid?: string;
  optionTestid?: string;
}

const Dropdown: FC<IProps> = (props) => {
  const {
    options = [],
    disabled,
    value,
    onChange,
    selectTestid,
    optionTestid,
  } = props;

  const [currentValue, setCurrentValue] = useState(value);

  const handleSelectChange = useCallback(
    (event: ChangeEvent<HTMLSelectElement>) => {
      const index = parseInt(event.currentTarget.value, 10);
      const label = options[index];
      setCurrentValue(index);

      if (onChange) {
        onChange(index, label);
      }
    },
    [options, onChange],
  );

  useEffect(() => setCurrentValue(value), [value]);

  return (
    <select
      className={styles.select}
      disabled={disabled}
      style={{
        backgroundImage: 'url("/icons/shevron.svg")',
      }}
      value={currentValue}
      onChange={handleSelectChange}
      data-testid={selectTestid}
    >
      {options.map((option, index) => (
        <option
          key={`option-${index}`}
          value={index}
          data-testid={optionTestid}
        >
          {option}
        </option>
      ))}
    </select>
  );
};

Dropdown.defaultProps = {
  value: 0,
  disabled: false,
  options: ['Option 1', 'Option 2'],
  onChange: undefined,
  selectTestid: 'select',
  optionTestid: 'option',
};

export default Dropdown;
