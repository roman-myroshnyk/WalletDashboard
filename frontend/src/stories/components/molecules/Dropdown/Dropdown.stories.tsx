import { Meta, Story } from '@storybook/react/types-6-0';
import Dropdown, { IProps } from '@/molecules/Dropdown';
import { CurrencyNames } from '@/consts/currency';

export default {
  title: 'molecules/Dropdown',
  component: Dropdown,
  argTypes: {
    options: {
      control: 'check',
      options: CurrencyNames,
    },
    value: {
      controls: 'number',
    },
    disabled: {
      controls: 'boolean',
    },
    onChange: {
      action: 'onChange: dropdown value changed',
    },
  },
} as Meta;

const Template:Story<IProps> = (args) => (
  <Dropdown {...args} />
);

export const Default = Template.bind({});

Default.args = {
  options: CurrencyNames,
  value: 1,
  disabled: false,
};
