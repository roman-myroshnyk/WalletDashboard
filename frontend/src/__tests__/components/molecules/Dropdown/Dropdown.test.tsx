import '.../../../mocks/framer-motion';
import { render, fireEvent } from '@testing-library/react';
import Dropdown from '@/molecules/Dropdown';

describe('molecules/Dropdown', () => {
  it('Dropdown should render', () => {
    const { container } = render(
      <Dropdown
        options={[
          'my option 1',
          'my option 2',
        ]}
        value={1}
      />,
    );
    expect(container).toMatchSnapshot();
  });
  it('Dropdown should trigger onChange handler', () => {
    const handleChange = jest.fn();
    const { getByTestId, getAllByTestId } = render(
      <Dropdown
        options={[
          'my option 1',
          'my option 2',
        ]}
        value={0}
        onChange={handleChange}
      />,
    );
    const select = getByTestId('select');
    const options = getAllByTestId('option') as HTMLOptionElement[];
    fireEvent.change(select, { target: { value: 1 } });
    expect(options[0].selected).toBeFalsy();
    expect(options[1].selected).toBeTruthy();
  });
  it('Dropdown should have disabled attribute', () => {
    const { getByTestId } = render(
      <Dropdown
        options={[
          'my option 1',
          'my option 2',
        ]}
        value={0}
        disabled
      />,
    );
    const select = getByTestId('select') as HTMLSelectElement;
    expect(select.disabled).toBeTruthy();
  });
});
