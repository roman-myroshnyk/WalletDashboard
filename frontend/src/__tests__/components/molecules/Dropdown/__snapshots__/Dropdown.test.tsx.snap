// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`molecules/Dropdown Dropdown should render 1`] = `
<div>
  <select
    class="select"
    data-testid="select"
    style="background-image: url(/icons/shevron.svg);"
  >
    <option
      data-testid="option"
      value="0"
    >
      my option 1
    </option>
    <option
      data-testid="option"
      value="1"
    >
      my option 2
    </option>
  </select>
</div>
`;
